### PR TITLE
Rewrite ResourceManager in C++

### PR DESCRIPTION
**Pull request by vabold**
This introduces an OOM when changing course in TTs. Too bad!

### CODE REVIEW

**Comment by riidefi** [comment]
Can you confirm said OOM was not already in `main`?

**Comment by vabold** [comment]
> Can you confirm said OOM was not already in `main`?

Yes.

File: payload/game/system/ResourceManager.hh [hunk 1]
```diff
@@ -1,25 +1,70 @@
...
+        enum class State {
+            Cleared = 0,
+            Loaded = 2,
+        };
+
+        void init();
+        void deinit();
+
+        void load(u32 courseId);
+        REPLACE void load(u32 courseId, bool splitScreen);
```
Comment (stblr) [inline payload/game/system/ResourceManager.hh:27]: In the C version the other overload was replaced.

Comment (vabold) [reply payload/game/system/ResourceManager.hh]: Resolved in latest commit.

File: payload/game/system/DvdArchive.cc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    m_archiveBuffer = other->m_archiveBuffer;
+    m_archiveSize = other->m_archiveSize;
+    m_archiveHeap = other->m_archiveHeap;
+    mount(m_archiveHeap);
+    other->m_archiveBuffer = nullptr;
+    other->m_archiveSize = 0;
+    other->m_archiveHeap = nullptr;
+    other->clearFile();
```
Comment (stblr) [inline payload/game/system/DvdArchive.cc:49]:
```suggestion
    other->clear();
```

File: payload/game/system/DvdArchive.hh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    void *REPLACED(getFile)(const char *path, size_t *size);
+    REPLACE void *getFile(const char *path, size_t *size);
+
+    void *buffer() const;
+    size_t size() const;
+    State state() const;
+
+private:
+    void clearFile();
```
Comment (stblr) [inline payload/game/system/DvdArchive.hh:32]:
```suggestion
    void clear();
```

File: symbols.txt [hunk 4]
```diff
@@ -478,15 +478,15 @@
 0x80511e7c CourseMap_computeFinishTime
 0x80511ec8 CheckPoint_computeFinishTime
 0x8051486c _ZmiRK4Vec3S1_
-0x80518dcc DvdArchive_mount
-0x80518e10 DvdArchive_load
-0x805190f0 DvdArchive_rip
+0x80518dcc _ZN6System10DvdArchive5mountEPN3EGG4HeapE
+0x80518e10 _ZN6System10DvdArchive4loadEPKcPN3EGG4HeapEbaS5_j
+0x805190f0 _ZN6System10DvdArchive3ripEPKcPN3EGG4HeapEa
 0x805192cc DvdArchive_clear
```
Comment (stblr) [inline symbols.txt:484]:
```suggestion
0x805192cc _ZN6System10DvdArchive5clearEv
```

**Review by stblr — COMMENTED** [review]
OOM fix just dropped.

**Comment by vabold** [comment]
> OOM fix just dropped.

Implemented.

**Review by stblr — APPROVED** [review]

**Pull request merged**